File: app/src/main/java/com/vls/mlkitqrscanner/barcodescanning/BarcodeScanningProcessor.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package com.vls.mlkitqrscanner.barcodescanning;

import com.google.android.gms.tasks.Task;
import com.google.firebase.ml.vision.FirebaseVision;
import com.google.firebase.ml.vision.barcode.FirebaseVisionBarcode;
import com.google.firebase.ml.vision.barcode.FirebaseVisionBarcodeDetector;
import com.google.firebase.ml.vision.common.FirebaseVisionImage;


import android.support.annotation.NonNull;
import android.util.Log;

import com.vls.mlkitqrscanner.FrameMetadata;
import com.vls.mlkitqrscanner.GraphicOverlay;
import com.vls.mlkitqrscanner.VisionProcessorBase;

import java.io.IOException;
import java.util.List;

/**
 * Barcode Detector Demo.
 */
public class BarcodeScanningProcessor extends VisionProcessorBase<List<FirebaseVisionBarcode>> {

    private static final String TAG = "BarcodeScanningProcessor";

    private final FirebaseVisionBarcodeDetector detector;
    private OnScanningListener scanningListener;

    public BarcodeScanningProcessor() {
        // Note that if you know which format of barcode your app is dealing with, detection will be
        // faster to specify the supported barcode formats one by one, e.g.
        // new FirebaseVisionBarcodeDetectorOptions.Builder()
        //     .setBarcodeFormats(irebaseVisionBarcode.FORMAT_QR_CODE)
        //     .build();
        detector = FirebaseVision.getInstance().getVisionBarcodeDetector();
    }

    public BarcodeScanningProcessor(OnScanningListener scanningListener) {
        this.scanningListener = scanningListener;
        detector = FirebaseVision.getInstance().getVisionBarcodeDetector();

    }

    @Override
    public void stop() {
        try {
            detector.close();
        } catch (IOException e) {
            Log.e(TAG, "Exception thrown while trying to close Barcode Detector: " + e);
        }
    }

    @Override
    protected Task<List<FirebaseVisionBarcode>> detectInImage(FirebaseVisionImage image) {
        return detector.detectInImage(image);
    }

    @Override
    protected void onSuccess(
            @NonNull List<FirebaseVisionBarcode> barcodes,
            @NonNull FrameMetadata frameMetadata,
            @NonNull GraphicOverlay graphicOverlay) {
        graphicOverlay.clear();

        if (scanningListener != null && barcodes.size() > 0) {
            scanningListener.getRawValue(barcodes.get(0).getRawValue());
            scanningListener = null;
        }

        /*for (int i = 0; i < barcodes.size(); ++i) {
            FirebaseVisionBarcode barcode = barcodes.get(i);
            BarcodeGraphic barcodeGraphic = new BarcodeGraphic(graphicOverlay, barcode);
            Log.e("VLS", barcode.getRawValue());
            graphicOverlay.add(barcodeGraphic);
        }*/
    }

    @Override
    protected void onFailure(@NonNull Exception e) {
        Log.e(TAG, "Barcode detection failed " + e);
    }

    public interface OnScanningListener {
        void getRawValue(String rawValue);
    }
}

File: app/src/main/java/com/vls/mlkitqrscanner/BarcodeScanningProcessor.kt
package com.vls.mlkitqrscanner

import com.google.android.gms.tasks.Task
import com.google.firebase.ml.vision.FirebaseVision
import com.google.firebase.ml.vision.barcode.FirebaseVisionBarcode
import com.google.firebase.ml.vision.barcode.FirebaseVisionBarcodeDetector
import com.google.firebase.ml.vision.common.FirebaseVisionImage
import android.util.Log


import java.io.IOException

class BarcodeScanningProcessor(scanningListener: OnScanningListener) : VisionProcessorBase<List<FirebaseVisionBarcode>>() {

    private val detector: FirebaseVisionBarcodeDetector = FirebaseVision.getInstance().visionBarcodeDetector
    private var scanningListener: OnScanningListener? = scanningListener

    // Note that if you know which format of barcode your app is dealing with, detection will be
    // faster to specify the supported barcode formats one by one, e.g.
    // new FirebaseVisionBarcodeDetectorOptions.Builder()
    //     .setBarcodeFormats(irebaseVisionBarcode.FORMAT_QR_CODE)
    //     .build();

    override fun stop() {
        try {
            detector.close()
        } catch (e: IOException) {
            Log.e(TAG, "Exception thrown while trying to close Barcode Detector: $e")
        }

    }

    override fun detectInImage(image: FirebaseVisionImage): Task<List<FirebaseVisionBarcode>> {
        return detector.detectInImage(image)
    }

    override fun onSuccess(
            barcodes: List<FirebaseVisionBarcode>,
            frameMetadata: FrameMetadata) {

        if (scanningListener != null && barcodes.isNotEmpty()) {
            scanningListener!!.getRawValue(barcodes[0].rawValue)
            scanningListener = null
        }

        /*for (int i = 0; i < barcodes.size(); ++i) {
            FirebaseVisionBarcode barcode = barcodes.get(i);
            BarcodeGraphic barcodeGraphic = new BarcodeGraphic(graphicOverlay, barcode);
            Log.e("VLS", barcode.getRawValue());
            graphicOverlay.add(barcodeGraphic);
        }*/
    }

    override fun onFailure(e: Exception) {
        Log.e(TAG, "Barcode detection failed $e")
    }

    interface OnScanningListener {
        fun getRawValue(rawValue: String?)
    }

    companion object {

        private const val TAG = "BarcodeScanningProcessor"
    }
}
